### ISSUE TITLE
[Bug] BD Installer hangs after "Downloading asar file"

### ISSUE THREAD

**Issue by CorneliusCornbread** [BetterDiscord/Installer#362]
### Before Reporting

- [X] I found no existing issues matching my bug
- [X] My issue is not caused by a plugin or theme

### Describe the Bug

Using the BetterDiscord app image on Fedora 39 it seems that every attempt at installing BetterDiscord leads to an installer hang. Running the app image through the terminal does not give any notable output, it just simply hangs. Sometimes it complains that it failed to download the package from the official website giving a 502 error, sometimes it just hangs while downloading the asar file.

Usual output before hanging
```
Downloading asar file
❌ Failed to download package from the official website
❌ Status code did not indicate success: 502
Falling back to GitHub...
```

### Expected Behavior

Better discord installer works

### Steps To Reproduce

1. Download app image
2. Run app image
3. Go through install process
4. See hang on install

### Operating System

Linux

### Discord Version

Stable

### BetterDiscord Version

v1.3.0

### Anything else?

Installer window is still responsive, it just stops doing anything.

**Comment by GH0STP4WZ** [comment]
I will open up a fedora VM momentarily to see if I can replicate the issue.
Where did you get your Discord package from?

**Comment by CorneliusCornbread** [comment]
> I will open up a fedora VM momentarily to see if I can replicate the issue. Where did you get your Discord package from?

I get it from the RPM fusion repos.

**Comment by Huderon** [comment]
I don't know nothing about nothing, but adding any kind of delay before the fallback request seemed to fix it for me. Was only able to reproduce on Fedora though so maybe that's just bunk?

**Comment by Vulpovile** [comment]
Same thing happens here but 
Downloading asar file
❌ Failed to download package from the official website
❌ Status code did not indicate success: 403
Falling back to GitHub...

**Comment by zerebos** [comment]
Did the fallback to github not work in your case?

**Comment by 999smokes** [comment]
same issue for me on windows and the fallback to github doesnt work

**Comment by Inve1951** [comment]
> same issue for me on windows and the fallback to github doesnt work

#368 is unrelated

**Comment by GH0STP4WZ** [comment]
> > I will open up a fedora VM momentarily to see if I can replicate the issue. Where did you get your Discord package from?
> 
> I get it from the RPM fusion repos.

I completely forgot to update the status on my testing for this issue. My Sincerest apologies. With the package mentioned installed, I was not able to replicate this issue. Perhaps I did not install the correct version of Fedora. My knowledge of what they produce is rather limited. However, From my knowledge of the installation process, it shouldn't be too hard to manually install BD either way.

You can try manually compiling the Asar or downloading it from the BetterDiscord repo. then you want to modify your discord installation and simply `require` the downloaded Asar file in the desktop_discord_core index.js

This wont fix your installer issue, but it will get BD running on your system.

**Comment by APT37** [comment]
```
Downloading asar file
❌ Failed to download package from the official website
❌ Client network socket disconnected before secure TLS connection was established
Falling back to GitHub...
❌ Failed to download package from https://api.github.com/repos/BetterDiscord/BetterDiscord/releases/assets/232165797
❌ Client network socket disconnected before secure TLS connection was established
```

All files are accessible in the browser, it's definitely a problem with the installer.

**Comment by GH0STP4WZ** [comment]
Have you tried downloading a new copy of the installer?

**Comment by CorneliusCornbread** [comment]
> Have you tried downloading a new copy of the installer?

Yes, here are the hashes of the file, tell me if the hashes are incorrect as it may point to a compromised distribution server.

MD5: `b041285c2f31c98983d2e28b8828db82`
SHA1: `b8f3cb95afaa1de53b8f1dcbd70b0210b47145da`
SHA256: `227e49e935a826c14e0c3c0c5ddd65320df7e35219243d8e25e6d5b60212c4fd`
SHA512: `0dd5a366832d5fba63552601a2c6b40e9ced35381002069876ae119ad85b4ee50fb990b55ec12756f91aa30fd3456825d1125a31294c09c039d991125a9c3cb5`

Funnily enough this was working again for the longest time for me but recently something changed.

**Comment by APT37** [comment]
The installer seems to have a problem with the redirect to GitHub. Changing the download URL to the redirect location and allowing status codes up to 302 fixed it for me.

That said, I don't know much about JavaScript and therefore won't commit a pull request, as this change could cause unexpected behavior I'm not aware of.

<details>
<summary>Broken</summary>

```js
try {
    const response = await downloadFile("https://betterdiscord.app/Download/betterdiscord.asar")
    const bdVersion = response.headers["x-bd-version"];
    if (200 <= response.statusCode && response.statusCode < 300) {
        log(`✅ Downloaded BetterDiscord version ${bdVersion} from the official website`);
        return response.body;
    }
    throw new Error(`Status code did not indicate success: ${response.statusCode}`);
}
```

</details>

<details>
<summary>Working</summary>

```js
try {
    const response = await downloadFile(
      "https://github.com/BetterDiscord/BetterDiscord/releases/latest/download/betterdiscord.asar"
    );
    if (200 <= response.statusCode && response.statusCode <= 302) {
        log(`✅ Downloaded BetterDiscord from GitHub`);
        return response.body;
    }
    throw new Error(`Status code did not indicate success: ${response.statusCode}`);
}
```

</details>

**Comment by CorneliusCornbread** [comment]
> The installer seems to have a problem with the redirect to GitHub. Changing the download URL to the redirect location and allowing status codes up to 302 fixed it for me.
> 
> That said, I don't know much about JavaScript and therefore won't commit a pull request, as this change could cause unexpected behavior I'm not aware of.
> 
> Broken:
> 
>     try {
>         const response = await downloadFile("https://betterdiscord.app/Download/betterdiscord.asar")
>         const bdVersion = response.headers["x-bd-version"];
>         if (200 <= response.statusCode && response.statusCode < 300) {
>             log(`✅ Downloaded BetterDiscord version ${bdVersion} from the official website`);
>             return response.body;
>         }
>         throw new Error(`Status code did not indicate success: ${response.statusCode}`);
>     }
> 
> Working:
> 
>     try {
>         const response = await downloadFile(
>           "https://github.com/BetterDiscord/BetterDiscord/releases/latest/download/betterdiscord.asar"
>         );
>         if (200 <= response.statusCode && response.statusCode <= 302) {
>             log(`✅ Downloaded BetterDiscord from GitHub`);
>             return response.body;
>         }
>         throw new Error(`Status code did not indicate success: ${response.statusCode}`);
>     }

Just tried this on my Fedora 41 machine, this didn't seem to change anything and resulted in the same issue.

**Comment by APT37** [comment]
Maybe try printing `response.location` and `response.body` to see what's happening. Are you by chance using a proxy?

**Comment by CorneliusCornbread** [comment]
> Maybe try printing `response.location` and `response.body` to see what's happening. Are you by chance using a proxy?

I was able to get it to install now, I recently upgraded AppImageLauncher to v3 so I wonder if that has something to do with it? If anyone else is having this problem you should try removing or upgrading to the new version to AppImageLauncher

**Comment by APT37** [comment]
Keeping your software up-to-date does help I guess. Since your problem is solved, you can close the issue.

**Comment by CorneliusCornbread** [comment]
Nevermind, after an update, despite using AppImageLauncher 3 this issue has resurfaced

Weirdly enough I found a work around?
1. Try repairing BD
2. Run through and hit yes to reinstall
3. Let it error
4. Hit the back button
5. Install
6. It works?

**Comment by APT37** [comment]
It's randomly working for me again.

**Comment by CorneliusCornbread** [comment]
So yeah it seems installing directly does not work for me, using the repair functionality seems to consistently work. I ran into this issue again, but was able to work around it only needing to repair instead of install.

Edit: ~~nevermind it didn't install at all~~ I got it with enough retrying my original workaround.

**Comment by APT37** [comment]
Still not working consistently. Used betterdiscordctl, which works without problems.

Now another issue arose, namely BetterDiscord not loading after opening Discord - as in it being stuck on showing the "BetterDiscord is loading..." message in the bottom right corner, indefinitely.

As the benefits don't outweigh the work I have to put into fixing these and other issues (or at least trying to), I've decided to not use BetterDiscord for the foreseeable future.

*I will not follow this or any other issue any further.*
